sql: refactor zone config setting code for multi-region

otan opened the pull request:
This commit refactors zone configuration code for multi-region to not
require shelling out to SQL and calling ALTER ... ZONE CONFIG. Instead,
zone configurations are done in-place. This reduces the number of SQL/KV
round trips and the need to look up the schema name of the database.

As an added bonus, added the requisite code that "merges"
non-multi-region based zone configurations. This means that, e.g.
setting gc.ttlseconds on a zone configuration carries over even after
altering localities.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59508)
<!-- Reviewable:end -->

aayushshah15 commented:
___
*[pkg/sql/region_util.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS92YijBrTAlaPph4XH:-MS92YijBrTAlaPph4XI:beancmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/148a99431917488ee9b58c77096199cb52bab98e/pkg/sql/region_util.go#L248)):*
> ```Go
> // mergeZoneConfigsForMultiRegion reassigns fields in `base` with the fields
> // that are multi-region related in `from`.
> func mergeZoneConfigsForMultiRegion(
> ```

drive by comment: are we sure this shallow-copy is safe? is there a reason we're not using the existing `CopyFromZone` here?


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (changes requested):
Left a few more comments since I'm in the area. Mainly, I think bypassing the zone config validation logic is a bad idea.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-:-MS9zXG61Y50p4KvtczC:b-grxug4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MS9qhmN0JDZdag2gGEa:bhvf1pu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
> ```Go
> 		base.CopyFromZone(*from, multiRegionZoneConfigFields)
> 	} else {
> 		base.CopyFromZone(*(zonepb.NewZoneConfig()), multiRegionZoneConfigFields)
> ```

Is this needed? `NewZoneConfig()` is empty so copying from it shouldn't do anything.  
___
*[pkg/sql/region_util.go, line 301 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9tnF60NRmDz52CPdB:-MS9tnF60NRmDz52CPdC:blx70ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L301)):*
> ```Go
> // applyZoneConfigForMultiRegionTable.
> type applyZoneConfigForMultiRegionTableOption struct {
> 	apply func(
> ```

For my understanding, is there a reason we're wrapping the function inside of the struct? Why not just typedef the function directly?
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MS9wLXN2x1qsoJSRk6J:b-c0ijgu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
> ```Go
> 	table catalog.TableDescriptor,
> 	opt applyZoneConfigForMultiRegionTableOption,
> ) error {
> ```

The way this is currently written, it will bypass all validation logic that we perform over zone configs. Is that intentional? I think it might be worth lifting out some logic from `applyZoneConfig` in `setZoneConfigNode.startExec()` to make sure we don't shoot ourselves in the foot. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajstorm)

___
*[pkg/sql/region_util.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS92YijBrTAlaPph4XH:-MS9X0mE4mm_8Cx7fChz:b-xbaixh) ([raw file](https://github.com/cockroachdb/cockroach/blob/148a99431917488ee9b58c77096199cb52bab98e/pkg/sql/region_util.go#L248)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

drive by comment: are we sure this shallow-copy is safe? is there a reason we're not using the existing `CopyFromZone` here?
</blockquote></details>

w.r.t. shallow copy: yes
w.r.t. CopyFromZone -- good find!
___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MSA37ogEZYwgaK2xwvJ:b-6od6vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Is this needed? `NewZoneConfig()` is empty so copying from it shouldn't do anything.  
</blockquote></details>

Yes, because NewZoneConfig sets these fields.

```
		InheritedConstraints:      true,
		InheritedLeasePreferences: true,
```
___
*[pkg/sql/region_util.go, line 301 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9tnF60NRmDz52CPdB:-MSA3C8-ChXw6Nyy3cs8:b-9vhu5c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L301)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

For my understanding, is there a reason we're wrapping the function inside of the struct? Why not just typedef the function directly?
</blockquote></details>

i did it this way so you can't form a function outside the package + i had some other non apply based options which turned out to not be needed, but it is private anyway so i guess so :P 
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSA4ppzClt0cDwsw1HL:bdrhijb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

The way this is currently written, it will bypass all validation logic that we perform over zone configs. Is that intentional? I think it might be worth lifting out some logic from `applyZoneConfig` in `setZoneConfigNode.startExec()` to make sure we don't shoot ourselves in the foot. 
</blockquote></details>

i've added the basic validation logic but avoided the ones that check live locality. how strongly do you feel about them?


<!-- Sent from Reviewable.io -->

ajstorm reviewed (approved):
Just nits from me, although I'll admit that I had some of the same comments that @aayushshah15 raised.  Will follow along...

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-:-MSAAI6G33HPRt6njFoA:b2uj8qt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality, line 582 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA0pEL2_DwWABLW_Tf:-MSA0pEL2_DwWABLW_Tg:b-48qb7q) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/ccl/logictestccl/testdata/logic_test/alter_table_locality#L582)):*
> ```
> # other fields.
> statement ok
> CREATE TABLE rbt_table_gc_ttl () LOCALITY REGIONAL BY TABLE IN "us-east-1";
> ```

Super Nit: extra semi-colon at the end
___
*[pkg/sql/alter_table_locality.go, line 98 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA1Qgp8EnRRvby55HA:-MSA1Qgp8EnRRvby55HB:bpzsuhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/898fe2a409d46226a3669b40edca841cdc77146d/pkg/sql/alter_table_locality.go#L98)):*
> ```Go
> 		return err
> 	}
> 	resolvedSchema, err := params.p.Descriptors().GetImmutableSchemaByID(
> ```

Heads-up - merge pain awaitith you.  TL;DR - this call is gone here, but was moved lower down.  You'll want to remove it from it's new location.
___
*[pkg/sql/alter_table_locality.go, line 160 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA2FMt0yQAXbkUoihf:-MSA2FMt0yQAXbkUoihg:b-1m1ryh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/alter_table_locality.go#L160)):*
> ```Go
> 	}
> 
> 	// Update the table's zone configuration.
> ```

Nit: Might want say "Update the zone configuration." here, as we're updating TableAndIndexes.
___
*[pkg/sql/region_util.go, line 210 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA2e967ZSyqmAm148K:-MSA2e967ZSyqmAm148L:b-8l935a) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L210)):*
> These fields

Which fields?

Also, might be good to define what you mean by "merged" here.  We overwrite the ones we're setting and leave the others alone?
___
*[pkg/sql/region_util.go, line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA4ljZ-xgvY9a6WOW7:-MSA4ljZ-xgvY9a6WOW8:boci0q4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L309)):*
> ```Go
> 
> // applyZoneConfigForMultiRegionTableOptionNewIndex applies table zone configs
> // for a newly added index  which requires partitioning of individual indexes.
> ```

Super Nit: double space
___
*[pkg/sql/region_util.go, line 355 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA75P94sDc-zyZJH8z:-MSA75P94sDc-zyZJH9-:bdeve0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L355)):*
> ```Go
> 		zc = mergeZoneConfigsForMultiRegion(zc, localityZoneConfig)
> 
> 		var hasNewSubzones bool
> ```

Nit: Why not?

`hasNewSubzones := table.IsLocalityRegionalByRow()`
`if hasNewSubzones {`
___
*[pkg/sql/region_util.go, line 432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA8eU2BTnsFuGNO7o7:-MSA8eU2BTnsFuGNO7o8:b-12nwo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L432)):*
> ```Go
> 			"delete-zone",
> 			txn,
> 			"DELETE FROM system.zones WHERE id = $1",
> ```

Why `DELETE FROM system.zones` instead of  `ALTER ... CONFIGURE ZONE DISCARD`?


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajstorm)

___
*[pkg/sql/region_util.go, line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS92YijBrTAlaPph4XH:-MSABGa0A_RY80lmcumJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/148a99431917488ee9b58c77096199cb52bab98e/pkg/sql/region_util.go#L248)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

w.r.t. shallow copy: yes
w.r.t. CopyFromZone -- good find!
</blockquote></details>

Done.
___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MSABH5HA1XzxySfovy-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Yes, because NewZoneConfig sets these fields.

```
		InheritedConstraints:      true,
		InheritedLeasePreferences: true,
```
</blockquote></details>

Done.
___
*[pkg/sql/region_util.go, line 301 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9tnF60NRmDz52CPdB:-MSABHp165JbWIEiJ0kL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L301)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i did it this way so you can't form a function outside the package + i had some other non apply based options which turned out to not be needed, but it is private anyway so i guess so :P 
</blockquote></details>

Done.
___
*[pkg/sql/region_util.go, line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA4ljZ-xgvY9a6WOW7:-MSABJbl5Jj5udjquC1U:b-cmkmxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L309)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Super Nit: double space
</blockquote></details>

the student has become the master
___
*[pkg/sql/region_util.go, line 432 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MSA8eU2BTnsFuGNO7o7:-MSABXbm8e3mxC_VyY5q:b-negl6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L432)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Why `DELETE FROM system.zones` instead of  `ALTER ... CONFIGURE ZONE DISCARD`?
</blockquote></details>

DISCARD requires the schema name, which we are trying to avoid
DISCARD deep down ends up doing this, so....


<!-- Sent from Reviewable.io -->

aayushshah15 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MSAUWmD2sDRG2wv08_K:beb2hqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Got it, do you think it'd be cleaner to return that from `zoneConfigFromTableLocalityConfig`  in the cases where it returns a `nil`?  We wouldn't need this merge method at all then.  
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSAFPhJ3H9uI30FWfpG:b-pwf679) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i've added the basic validation logic but avoided the ones that check live locality. how strongly do you feel about them?
</blockquote></details>

I'd strongly prefer it if we didn't have two (partially overlapping) sets of validation logic in two separate places, gratuitously. In other words, we should have all hydrated zone configs be validated by calling out to a single method that houses all checks. This is so that future checks can be added in one place and we can assume that all zone configs inside `system.zones` satisfy the same set of invariants. 


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aayushshah15 and @ajstorm)

___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MSAzmh71GJx1VfhRSsD:b7q5qv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Got it, do you think it'd be cleaner to return that from `zoneConfigFromTableLocalityConfig`  in the cases where it returns a `nil`?  We wouldn't need this merge method at all then.  
</blockquote></details>

I think I prefer them separate because it seems more composable and separate in responsibility, unless you have strong feelings about it.
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSAzz5c-LVjeYBO8c41:b-hc2xjc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

I'd strongly prefer it if we didn't have two (partially overlapping) sets of validation logic in two separate places, gratuitously. In other words, we should have all hydrated zone configs be validated by calling out to a single method that houses all checks. This is so that future checks can be added in one place and we can assume that all zone configs inside `system.zones` satisfy the same set of invariants. 
</blockquote></details>

which bits of validation are you looking for exactly? I have done the Validate() call below -- seems fairly easily compartmentalised.

there are other cases outside set_zone_config which does direct modification of zone configs (e.g. alter_primary_key) -- if you want a bigger "overhaul" in general of validation i can make an issue for it. from a brief glance, the `Validate` function has separate steps than `validateNoRepeatKeysInZone`, which they should be combined -- i can file a separate issue for that.

(ideally there's a trigger on the system.zones table for validation, but that's a different matter)


<!-- Sent from Reviewable.io -->

otan committed:
sql: refactor zone config setting code for multi-region

This commit refactors zone configuration code for multi-region to not
require shelling out to SQL and calling ALTER ... ZONE CONFIG. Instead,
zone configurations are done in-place. This reduces the number of SQL/KV
round trips and the need to look up the schema name of the database.

As an added bonus, added the requisite code that "merges"
non-multi-region based zone configurations. This means that, e.g.
setting gc.ttlseconds on a zone configuration carries over even after
altering localities.

Release note: None

aayushshah15 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> TY for entertaining my concerns 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-:-MSBTnSVEvwomLGWiwC6:b66fvq6)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 260 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9qhmN0JDZdag2gGE_:-MSBQgLEAgKVz4DVI-Ol:b1jxt0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L260)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

I think I prefer them separate because it seems more composable and separate in responsibility, unless you have strong feelings about it.
</blockquote></details>

I do not :+1: 
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSBRunz-d8UBg2S3yj0:bqspz32) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
> there are other cases outside set_zone_config which does direct modification of zone configs (e.g. alter_primary_key)

I suppose in `alter_primary_key` its fair to skip validation since we're simply migrating existing zone configs to point to one index from another.

Making an issue for this stuff sounds ok to me, if you wouldn't mind?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSBj9__0qqeQKnCb5PH:bv3cdle) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

> there are other cases outside set_zone_config which does direct modification of zone configs (e.g. alter_primary_key)

I suppose in `alter_primary_key` its fair to skip validation since we're simply migrating existing zone configs to point to one index from another.

Making an issue for this stuff sounds ok to me, if you wouldn't mind?
</blockquote></details>

Of course, #59572.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59508)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @ajstorm, and @otan)

___
*[pkg/sql/region_util.go, line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59508#-MS9wLXN2x1qsoJSRk6I:-MSBj9__0qqeQKnCb5PH:bv3cdle) ([raw file](https://github.com/cockroachdb/cockroach/blob/f65751d3b5fd6bb074bec9790e6bb279816908bb/pkg/sql/region_util.go#L392)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

> there are other cases outside set_zone_config which does direct modification of zone configs (e.g. alter_primary_key)

I suppose in `alter_primary_key` its fair to skip validation since we're simply migrating existing zone configs to point to one index from another.

Making an issue for this stuff sounds ok to me, if you wouldn't mind?
</blockquote></details>

Of course, #59572.


<!-- Sent from Reviewable.io -->

otan commented:
thxxxx

bors r=aayushshah15
